fix(typescript-estree): only run projectService setHostConfiguration when needed

higherorderfunctor opened the pull request:
## PR Checklist

- [x] Addresses an existing open issue: fixes #9312
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

~~Moves `ProjectService.setHostConfiguration` from `useProgramFromProjectService` where it is called on every source to `createProjectService` where it is only called once per `ProjectService`.~~ See comments regarding updated implementation.

`ProjectService.setHostConfiguration` [reloads](https://github.com/microsoft/TypeScript/blob/27bcd4cb5a98bce46c9cdd749752703ead021a4b/src/server/editorServices.ts#L3458) the whole project.  The way `vue-eslint-parser` sends template fragments causes a reload to start and then get [canceled](https://github.com/microsoft/TypeScript/blob/27bcd4cb5a98bce46c9cdd749752703ead021a4b/src/server/editorServices.ts#L3504) for _n - 1_ template fragments per file.

This fix along with https://github.com/typescript-eslint/typescript-eslint/issues/9223 if using `eslint-plugin-import`/`eslint-plugin-import-x` vastly improves performance.

~~I am open to suggestions on improving the signature of `createProjectService`.  I appended as a new 3rd argument called `parseSettings` which is an object to allow for additional settings to be added in the future if needed.  I could see this being merged with the second argument, but I was not sure of the implications in making a breaking change to the existing signature.~~

JoshuaKGoldberg commented on packages/typescript-estree/src/create-program/createProjectService.ts:
```diff
@@ -22,9 +22,14 @@ export interface ProjectServiceSettings {
   service: TypeScriptProjectService;
 }
 
+export interface ProjectServiceParseSettings {
+  extraFileExtensions?: string[];
+}
+
 export function createProjectService(
   optionsRaw: boolean | ProjectServiceOptions | undefined,
   jsDocParsingMode: ts.JSDocParsingMode | undefined,
+  parseSettings?: ProjectServiceParseSettings,
```
[Refactor] How about moving `jsDocParsingMode` into `ProjectServiceParseSettings`? They're both about parsing, after all:

```suggestion
export interface ProjectServiceParseSettings {
  extraFileExtensions?: string[];
  jsDocParsingMode?: ts.JSDocParsingMode;
}

export function createProjectService(
  optionsRaw: boolean | ProjectServiceOptions | undefined,
  parseSettings?: ProjectServiceParseSettings,
```

higherorderfunctor replied:
I had to refactor to support an ESLint configuration that changes `extraFileExtensions` in the same linting session.  As such, I've moved the logic back to `packages/typescript-estree/src/useProgramFromProjectService.ts`.

JoshuaKGoldberg reviewed (changes requested):
Tentatively looks lovely to me, with the refactor suggestion factored in. I'll take a deeper look once I have a chance to play with a repro.

cc @jakebailey as an FYI for this & the issue 🙂

higherorderfunctor commented:
Thanks for the initial review.  I did discover one bug that is fixed in my branch with a collection of changes.  If a user changes extensions from the initial ones, they do not get updated.  I'll carry the fix in and add a test.

higherorderfunctor commented:
Taking out of draft.  Notable updates:

* Discovered a bug when changing `extraFileExtensions` in the same linting session.
  * Added positive and negative test cases for this situation.
* Moved the logic back to `packages/typescript-estree/src/useProgramFromProjectService.ts`.
* The approach now caches the last set `extraFileExtensions` and only calls `setHostConfiguration` when there is a difference.
  * I'm caching directly on the `ProjectService` with `__extra_file_extensions` so it gets garbage collected with `clearTSServerProjectService`.  If you prefer a proper cleanup with it's own function, I can do so, but it may take a few days to find availability.
  * Using sets yielded the best performance when testing hence the `symmetricDifference` logic.  I did not exhaustively look at other options, just some tests iterating with for-loops.  Let me know if you want a different approach.
* I've added documentation to the site explaining how to minimize project reloads by setting `extraFileExtensions` only once.

This is part of a larger set of changes discussed on https://github.com/typescript-eslint/typescript-eslint/issues/9312#issuecomment-2166787001 and https://github.com/typescript-eslint/typescript-eslint/issues/8835#issuecomment-2164423027.

- Fixes `extends` not getting processed by the default config used for `projectService` inferred projects. https://github.com/typescript-eslint/typescript-eslint/pull/9306
- Adds tsserver logging. https://github.com/typescript-eslint/typescript-eslint/pull/9337
- Minimizes how many times `setHostConfiguration` is called. https://github.com/typescript-eslint/typescript-eslint/pull/9336
- Adds support to use watches with a Trie lookup (without actual FS watching) to update tsserver with changes (e.g., file created) to fix https://github.com/typescript-eslint/typescript-eslint/issues/8835.
- Caches open files with tsserver and closes them using an LRU based on access order.
- Sends incremental diffs instead of re-opening files if already open.

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,6 +79,86 @@ module.exports = {
 
+## Changes to `extraFileExtensions` with `projectService` in your ESLint options
+
+Using different [`extraFileExtensions`](../packages/Parser.mdx#extrafileextensions) for different files with [`projectService`](../packages/Parser.mdx#projectservice) will cause a full project reload.
```
[Docs]

> will cause a full project reload

This is a little vague out of context. Can you be a bit more specific? In what scenario(s) does it cause the reload? How does that impact code? How would a user observe this?

higherorderfunctor replied:
Changing `extraFileExtensions` with tsservers `setHostConfiguration` causes tsserver to initiate a project reload to update the project structure with the extension changes: https://github.com/microsoft/TypeScript/blob/b1c52c53cc6c8cf35b19accfc3e29916489c821e/src/server/editorServices.ts#L3781

Tsserver does not appear to do any checks if the old file extensions match the new file extensions.

The current implementation of `typescript-estree` updates the file extension for every file when `extraFileExtensions` is set.  This compounds with plugins like `vue-eslint-parser` where each template fragment triggers this.

This PR caches the last set `extraFileExtensions` to minimize these project reloads.

Users can further optimize, which is why I added it to the docs, by making sure `extraFileExtensions` is only set once.

I can merge in main now that the logging PR was accepted to expose tsserver logs and provide some examples.

A user could look for this log message as an example: https://github.com/microsoft/TypeScript/blob/b1c52c53cc6c8cf35b19accfc3e29916489c821e/src/server/editorServices.ts#L3825

JoshuaKGoldberg replied:
Oh, just clarifying, I meant those questions as rhetorical. They're things that the docs should answer. I'm requesting additions to the docs. Sorry for not being clear 😅

higherorderfunctor replied:
Additions have been added.  Let me know if additional detail is needed in any area.

I added my [own comment](https://github.com/typescript-eslint/typescript-eslint/pull/9336#discussion_r1643353701) to point out a creative liberty taken to help with readability.

I can revert for accuracy if preferred.

JoshuaKGoldberg replied:
These additions are fantastic! I hadn't thought of adding an FAQ article, but it makes a ton of sense. Strong +1 😄

jakebailey commented on packages/typescript-estree/src/useProgramFromProjectService.ts:
```diff
@@ -13,6 +13,46 @@ const log = debug(
+const difference = <T>(self: Set<T>, other: Set<T>): Set<T> =>
+  new Set([...self].filter(elem => !other.has(elem)));
+const symmetricDifference = <T>(self: Set<T>, other: Set<T>): Set<T> =>
+  union(difference(self, other), difference(other, self));
```
Does it have to be even that complicated? It seems like overkill to optimize even for order, in which case you may as well just use [`util.isDeepStrictEqual`](https://nodejs.org/api/util.html#utilisdeepstrictequalval1-val2) or something.

higherorderfunctor replied:
Average of 100 runs on the `AllkindsICP` package from the issue: before=0:28.36 after=0:28.96.

Not enough of a difference to matter.  Moving forward with `isDeepStrictEqual`.

higherorderfunctor replied:
These changes have been made!

higherorderfunctor commented on docs/troubleshooting/Performance.mdx:
````diff
@@ -79,7 +79,113 @@ module.exports = {
+Project reloads can be observed using the [debug environment variable](../packages/typescript-estree/#debugging): `DEBUG='typescript-eslint:typescript-estree:*'`.
+
+```
+typescript-estree:useProgramFromProjectService Updating extra file extensions: before=[]: after=[ '.vue' ]
````
I trimmed these log messages on the left so they would (mostly) fit horizontally on screen.  If accuracy is preferred, I can undo this.

```typescript-estree:useProgramFromProjectService```

vs

```2024-06-17T16:58:28.942Z typescript-eslint:typescript-estree:useProgramFromProjectService```

JoshuaKGoldberg replied:
Yeah +1 this is preferable IMO. I don't think users will be upset at not having to read through arbitrary timestamps. 😄

JoshuaKGoldberg replied:
That being said, could you move this debug information to the bottom? I think it'd be easier for folks to have the order be:

1. Introduction to this idea (don't change `extraFileExtensions` between files)
2. Why this is the case (full project reloads)
3. Ancillary info, like debugging tips

higherorderfunctor commented on packages/typescript-estree/src/useProgramFromProjectService.ts:
```diff
@@ -13,6 +13,46 @@ const log = debug(
+
+const updateExtraFileExtensions = (
+  service: ts.server.ProjectService & {
+    __extra_file_extensions?: Set<string>;
```
Updated to `WeakMap`.  Thanks!

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,7 +79,120 @@ module.exports = {
+
+As files are received, if `extraFileExtensions` differs from the previous file,
+the TypeScript server is updated to match your ESLint options.
+The TypeScript server incrementally updates the project structure by creating an execution pipeline.
```
[Docs] I think we can get away without mentioning the concept of the execution pipeline...

```suggestion
```

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
````diff
@@ -79,7 +79,120 @@ module.exports = {
+the TypeScript server is updated to match your ESLint options.
+The TypeScript server incrementally updates the project structure by creating an execution pipeline.
+If the ordering of files received from ESLint happens to frequently alternate the value of `extraFileExtensions`,
+a project reload may occur before the last pipeline completes.
+In turn, this can lead to cascading performance issues.
+
+```
+typescript-estree:tsserver:info Scheduled: /path/to/tsconfig.src.json, Cancelled earlier one +0ms
+typescript-estree:tsserver:info Scheduled: *ensureProjectForOpenFiles*, Cancelled earlier one +0ms
+```
+
+The impact of setting `extraFileExtensions` per file type in a project may be negligible depending on the order files are received.
+That said, complex ESLint options, cached linting results altering which files are evaluated,
````
...and trimming out further details that enforce the idea of _"don't let this change"_:

```suggestion
```

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,7 +79,120 @@ module.exports = {
 See [Glob pattern in parser's option "project" slows down linting](https://github.com/typescript-eslint/typescript-eslint/issues/2611) for more details.
 
-## The `indent` / `@typescript-eslint/indent` rules
+## Changes to `extraFileExtensions` with `projectService` in your ESLint options
```
[Docs] A bit more trim:

```suggestion
## Changes to `extraFileExtensions` with `projectService`
```

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,7 +79,120 @@ module.exports = {
+may all be sources of unintended project reloads.
+The optimal approach to avoid project reloads altogether is to use the same `extraFileExtensions` across `parserOptions`.
+
+<Tabs groupId="eslint-config">
```
[Praise] These diffs are very helpful 🙂 nice!

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,7 +79,120 @@ module.exports = {
+</TabItem>
+</Tabs>
+
+## The `indent` / `@typescripst-eslint/indent` rules
```
oopity

```suggestion
## The `indent` / `@typescript-eslint/indent` rules
```

JoshuaKGoldberg reviewed (approved):
Looks great to me, thanks! 🙌 

Mostly touching up on the docs FAQs. I think they're pretty close, just need some reordering.

higherorderfunctor commented:
Made the reordering and trimming changes requested.  I also updated the links with the title change in the performance section.

Let me know if any further edits are needed!

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,6 +79,107 @@ module.exports = {
 
+## Changes to `extraFileExtensions` with `projectService`
+
+Changes to [`extraFileExtensions`](../packages/Parser.mdx#extrafileextensions) between files in the same project with
```
[Docs] Nice tip I've picked up, using different wording in the header vs. the body. The two tend to have different kind of phrasing, and this body text is a bit more precise.

```suggestion
Using a different `extraFileExtensions`](../packages/Parser.mdx#extrafileextensions) between files in the same project with
```

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,6 +79,107 @@ module.exports = {
+## Changes to `extraFileExtensions` with `projectService`
+
+Changes to [`extraFileExtensions`](../packages/Parser.mdx#extrafileextensions) between files in the same project with
+the [`projectService`](../packages/Parser.mdx#projectservice) option is likely to cause more overhead than expected updating the project.
```
[Docs] For succinctness:

```suggestion
the [`projectService`](../packages/Parser.mdx#projectservice) option may cause performance degredations.
```

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,6 +79,107 @@ module.exports = {
+
+Changes to [`extraFileExtensions`](../packages/Parser.mdx#extrafileextensions) between files in the same project with
+the [`projectService`](../packages/Parser.mdx#projectservice) option is likely to cause more overhead than expected updating the project.
+The optimal approach is to use the same `extraFileExtensions` value across all files in the project.
```
[Docs] Redundant with the previous sentence, I think.

```suggestion
```

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,6 +79,107 @@ module.exports = {
+the [`projectService`](../packages/Parser.mdx#projectservice) option is likely to cause more overhead than expected updating the project.
+The optimal approach is to use the same `extraFileExtensions` value across all files in the project.
+For every file linted, we update the `projectService` whenever `extraFileExtensions` changes to match your ESLint options.
+This causes the underlying TypeScript server to perform a full project reload to account for new and removed files from the project.
```
[Docs] Not necessary to mention for this, I think:

```suggestion
This causes the underlying TypeScript server to perform a full project reload.
```

JoshuaKGoldberg commented on docs/troubleshooting/Performance.mdx:
```diff
@@ -79,6 +79,107 @@ module.exports = {
+Changes to [`extraFileExtensions`](../packages/Parser.mdx#extrafileextensions) between files in the same project with
+the [`projectService`](../packages/Parser.mdx#projectservice) option is likely to cause more overhead than expected updating the project.
+The optimal approach is to use the same `extraFileExtensions` value across all files in the project.
+For every file linted, we update the `projectService` whenever `extraFileExtensions` changes to match your ESLint options.
```
[Docs] Redundant:

```suggestion
For every file linted, we update the `projectService` whenever `extraFileExtensions` changes.
```

JoshuaKGoldberg reviewed (approved):
Lovely, thanks for this! 🙌 

I just have docs nitpicks left over. We'll make sure this is merged before our Monday release in two days. 🚀 thanks a bunch @higherorderfunctor!

higherorderfunctor commented:
All review suggestions have been committed.  Thanks again!

higherorderfunctor commented:
Inspecting pipeline errors and some of the links aren't rendering locally.  Inspecting now.

higherorderfunctor commented:
Fixes made.  So long as the pipeline succeeds, we should be good to merge!

JoshuaKGoldberg reviewed (approved):
🔥 ! Very excited to get this released, thank you! A bit of a relief to fix such a disruptive performance issue for Vue users.

I'll merge this in now, merge #9024 into `main`, and then merge `main` into `v8` - so all the FAQs including this one will be nicely organized.
